File: tests/typescript/unit/sdk/context.test.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Unit tests for SDK context propagation: bind/current, capture, overrides, and envelope round-trips.

import { describe, it, expect } from 'vitest'
import {
  bind,
  current,
  captureContext,
  withOverrides,
  toEnvelope,
  fromEnvelope,
  describeAuthority,
  type CaracalContext,
} from '../../../../packages/sdk/ts/src/context.js'

function ctx(overrides: Partial<CaracalContext> = {}): CaracalContext {
  return {
    subjectToken: 'tok',
    zoneId: 'zone-1',
    clientId: 'app-1',
    agentSessionId: 'agent-1',
    sessionId: 'sess-1',
    traceId: 'trace-1',
    hop: 0,
    ...overrides,
  }
}

describe('bind and current', () => {
  it('exposes the bound context inside the callback and clears it outside', async () => {
    expect(current()).toBeUndefined()
    const seen = await bind(ctx(), async () => current())
    expect(seen?.agentSessionId).toBe('agent-1')
    expect(current()).toBeUndefined()
  })
})

describe('captureContext', () => {
  it('returns undefined with no active context', () => {
    expect(captureContext()).toBeUndefined()
  })

  it('returns a detached copy of the active context', async () => {
    await bind(ctx(), async () => {
      const snap = captureContext()
      expect(snap).toEqual(current())
      expect(snap).not.toBe(current())
    })
  })
})

describe('withOverrides', () => {
  it('throws when no base context exists', () => {
    expect(() => withOverrides({ hop: 9 }, () => undefined)).toThrow(/requires an existing/)
  })

  it('merges overrides onto the base context', async () => {
    await bind(ctx(), async () => {
      const merged = await withOverrides({ hop: 5, clientId: 'app-2' }, async () => current())
      expect(merged?.hop).toBe(5)
      expect(merged?.clientId).toBe('app-2')
      expect(merged?.zoneId).toBe('zone-1')
    })
  })
})

describe('envelope round-trip', () => {
  it('serializes and restores a context via the envelope', () => {
    const original = ctx({ delegationEdgeId: 'edge-1', parentEdgeId: 'edge-0', hop: 2 })
    const env = toEnvelope(original)
    const restored = fromEnvelope(env, { zoneId: 'zone-1', clientId: 'app-1' })
    expect(restored).toMatchObject({
      subjectToken: 'tok',
      agentSessionId: 'agent-1',
      delegationEdgeId: 'edge-1',
      parentEdgeId: 'edge-0',
      hop: 2,
    })
  })

  it('rejects an envelope without a subject token', () => {
    expect(() => fromEnvelope({ hop: 0 } as never, { zoneId: 'z', clientId: 'c' }))
      .toThrow(/missing subject token/)
  })
})

describe('describeAuthority', () => {
  it('returns undefined without a context', () => {
    expect(describeAuthority(undefined)).toBeUndefined()
  })

  it('builds the full authority chain in order', () => {
    const summary = describeAuthority(ctx({
      delegationEdgeId: 'edge-1',
      parentEdgeId: 'edge-0',
      hop: 3,
    }))
    expect(summary?.chain).toEqual([
      'authority:sess-1',
      'agent-run:agent-1',
      'parent-delegated-permission:edge-0',
      'delegated-permission:edge-1',
    ])
    expect(summary).toMatchObject({ zoneId: 'zone-1', applicationId: 'app-1', hop: 3 })
  })

  it('omits chain segments for absent identifiers', () => {
    const summary = describeAuthority(ctx({ agentSessionId: undefined, sessionId: undefined }))
    expect(summary?.chain).toEqual([])
  })
})

File: tests/source/go/packages/sdk/go/context_test.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Unit tests for CaracalContext bind/current and envelope projection.

package sdk_test

import (
	"context"
	"strings"
	"testing"

	sdk "github.com/garudex-labs/caracal/packages/sdk/go"
)

func TestBindCurrentRoundTrip(t *testing.T) {
	c := sdk.CaracalContext{
		SubjectToken:     "tok",
		ZoneID:           "z1",
		ClientID:         "app1",
		AgentSessionID:   "sess",
		DelegationEdgeID: "edge",
		SessionID:        "sid",
		Hop:              2,
	}
	ctx := sdk.Bind(context.Background(), c)
	got, ok := sdk.Current(ctx)
	if !ok {
		t.Fatal("Current must return true after Bind")
	}
	if got.SubjectToken != c.SubjectToken {
		t.Errorf("SubjectToken: %q vs %q", got.SubjectToken, c.SubjectToken)
	}
	if got.ZoneID != c.ZoneID {
		t.Errorf("ZoneID: %q vs %q", got.ZoneID, c.ZoneID)
	}
	if got.Hop != c.Hop {
		t.Errorf("Hop: %d vs %d", got.Hop, c.Hop)
	}
}

func TestCurrentOnFreshContext(t *testing.T) {
	_, ok := sdk.Current(context.Background())
	if ok {
		t.Error("Current must return false on a context with no Bind")
	}
}

func TestBindDoesNotMutateParent(t *testing.T) {
	parent := context.Background()
	sdk.Bind(parent, sdk.CaracalContext{SubjectToken: "tok"})
	_, ok := sdk.Current(parent)
	if ok {
		t.Error("Bind must not modify the parent context")
	}
}

func TestFromEnvelopeFullFields(t *testing.T) {
	env := sdk.Envelope{
		SubjectToken:     "tok",
		AgentSessionID:   "sess",
		DelegationEdgeID: "edge",
		ParentEdgeID:     "parent",
		SessionID:        "sid",
		TraceID:          "0123456789abcdef0123456789abcdef",
		Hop:              4,
	}
	c, err := sdk.FromEnvelope(env, "zone1", "app1")
	if err != nil {
		t.Fatal(err)
	}
	if c.SubjectToken != "tok" {
		t.Errorf("SubjectToken: %q", c.SubjectToken)
	}
	if c.ZoneID != "zone1" {
		t.Errorf("ZoneID: %q", c.ZoneID)
	}
	if c.AgentSessionID != "sess" {
		t.Errorf("AgentSessionID: %q", c.AgentSessionID)
	}
	if c.SessionID != "sid" {
		t.Errorf("SessionID: %q", c.SessionID)
	}
	if c.Hop != 4 {
		t.Errorf("Hop: %d", c.Hop)
	}
}

func TestFromEnvelopeMissingSubjectTokenErrors(t *testing.T) {
	_, err := sdk.FromEnvelope(sdk.Envelope{}, "z", "a")
	if err == nil {
		t.Fatal("expected error for missing subject token")
	}
}

func TestToEnvelopeRoundTrip(t *testing.T) {
	c := sdk.CaracalContext{
		SubjectToken:     "tok",
		AgentSessionID:   "sess",
		DelegationEdgeID: "edge",
		ParentEdgeID:     "parent",
		SessionID:        "sid",
		TraceID:          "0123456789abcdef0123456789abcdef",
		Hop:              2,
	}
	env := sdk.ToEnvelope(c)
	if env.SubjectToken != c.SubjectToken {
		t.Errorf("SubjectToken: %q vs %q", env.SubjectToken, c.SubjectToken)
	}
	if env.AgentSessionID != c.AgentSessionID {
		t.Errorf("AgentSessionID: %q vs %q", env.AgentSessionID, c.AgentSessionID)
	}
	if env.TraceID != c.TraceID {
		t.Errorf("TraceID: %q vs %q", env.TraceID, c.TraceID)
	}
	if env.SessionID != c.SessionID {
		t.Errorf("SessionID: %q vs %q", env.SessionID, c.SessionID)
	}
	if env.Hop != c.Hop {
		t.Errorf("Hop: %d vs %d", env.Hop, c.Hop)
	}
}

func TestToEnvelopeFromEnvelopeRoundTrip(t *testing.T) {
	orig := sdk.CaracalContext{
		SubjectToken:     "tok",
		ZoneID:           "z",
		ClientID:         "app",
		AgentSessionID:   "sess",
		DelegationEdgeID: "edge",
		SessionID:        "sid",
		Hop:              1,
	}
	env := sdk.ToEnvelope(orig)
	restored, err := sdk.FromEnvelope(env, orig.ZoneID, orig.ClientID)
	if err != nil {
		t.Fatal(err)
	}
	if restored.SubjectToken != orig.SubjectToken {
		t.Errorf("SubjectToken mismatch")
	}
	if restored.AgentSessionID != orig.AgentSessionID {
		t.Errorf("AgentSessionID mismatch")
	}
	if restored.SessionID != orig.SessionID {
		t.Errorf("SessionID mismatch")
	}
}

func TestDescribeAuthorityRedactsSubjectToken(t *testing.T) {
	ctx := sdk.Bind(context.Background(), sdk.CaracalContext{
		SubjectToken:     "tok",
		ZoneID:           "z",
		ClientID:         "app",
		SessionID:        "sid",
		AgentSessionID:   "agent",
		DelegationEdgeID: "edge",
		Hop:              2,
	})
	summary, ok := sdk.DescribeAuthority(ctx)
	if !ok {
		t.Fatal("DescribeAuthority must return a summary for a bound context")
	}
	if summary.ApplicationID != "app" || summary.AuthoritySessionID != "sid" || summary.AgentRunID != "agent" {
		t.Fatalf("unexpected summary: %#v", summary)
	}
	got := strings.Join(summary.Chain, ">")
	want := "authority:sid>agent-run:agent>delegated-permission:edge"
	if got != want {
		t.Fatalf("chain = %q, want %q", got, want)
	}
}
